### PR TITLE
Implemented regex and validation helper functions and fixed some bugs

### PR DESCRIPTION
**Pull request by AlmirKadric**
Added regex helper function which turns a comma separated numerical range (i.e. 1,2,5-10) into a regular expression for matching.

Also added a some validation helper functions which can test to see if specific arguments are provided etc.

Lastly fixed an issue where bashr.sh Setup function fails exit on a non zero code when it fails. This is an issue when determining the success/failure of a script based on exit codes.

### CODE REVIEW

File: regex.sh [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+# Unit test function for RangeToRegex. Uses TestRangeToRegex to test multiple
+# known edge cases. Should be added to overtime as edge cases arrise or are
+# thought of.
+#
+# Usage: RangeToRegexUnitTest [<ranges>]
+#
+# Arguments:
+#     ranges: Optional comma separated list of ranges. i.e. "1,5,9,10-20,400-1000"
+##
+function RangeToRegexUnitTest() {
```
Comment (stelcheck) [inline regex.sh]: Its good that you create unit tests, but they should be separate from the actual code. Please create a /test dir, and put your unit tests in a separate file(s).

example:

/test/regex/01_range2regex.sh
/test/regex/02_someotherregextest.sh
etc

Edit:

You should not have to feed manually data to a unit test: the unit test should contain the data. So maybe something like

/test/regex/01_range2regex_singleValue.sh
/test/regex/02_range2regex_range.sh
/test/regex/03_range2regex_list.sh
/test/regex/02_range2regex_rangeAndList.sh

File: regex.sh [hunk 2]
```diff
@@ -0,0 +1,453 @@
+#!/bin/bash
+
+##
+# title           : regex.sh
+# description     : This script provides helper functions for
+#                   creating or manipulating regular expressions.
```
Comment (stelcheck) [inline regex.sh:6]: It does not look like you support letters or words... should it be called regex.sh then?

File: validation.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+##
+# Takes in a space separated list of command which should be expected. If anyone
+# one of the listed commands are found to not exist, an error will thrown for
+# that command. And once all commands are processed, if any were found missing
+# function will exit with status code of 1.
+# 
+# Commands can be scripts, system binaries etc. found in $PATH or functions
+#
+# Usage: EnsureCommandsAvailable <commands>
```
Comment (stelcheck) [inline validation.sh]: this should be 

Usage: EnsureCommandsAvailable command1 [command2] [...]

Thus making it clear that they are separate arguments

File: validation.sh [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+##
+# Function which takes in a list of arguments and allowed arguments and then
+# determines if there are any foreign arguments. If none are found it will then
+# store the argument value pairs into variables with argument label and values.
+#
+# Usage: ParseArguments <allowed> <arguments>
+#
+# Arguments:
+#     allowed: space separated list of allowed arguments. Anything not in this
```
Comment (stelcheck) [inline validation.sh]: might be better to make it comma-seperated? This way, you dont have to quote your list, hence simplifying input.

File: validation.sh [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#              list will be considered foreign and will cause the function to
+#              exit with error code 1.
+#
+#   arguments: list of arguments passed to your application (usually will be $@)
```
Comment (stelcheck) [inline validation.sh:94]: might be better to make it comma-seperated? This way, you dont have to quote your list, hence simplifying input.

**Comment by stelcheck** [comment]
in regex.sh, arguments are not checked for proper input.

Examples

mt@kodachi:~/libbash (libbash::master)$ TestRangeToRegex 1-2,4 1-5
-bash: [: 2,4: integer expression expected
Usage: TestFalsePositives <range> <test from> <test to>
mt@kodachi:~/libbash (libbash::master)$ TestRangeToRegex 1-2,4 1 5
-bash: [: 2,4: integer expression expected
seq: invalid floating point argument: 2,4
Try `seq --help' for more information.

File: regex.sh [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+    then
+        echo "Usage: TestFalsePositives <range> <test from> <test to>";
+        return 1;
+    fi
+
+    # Get regex
+    regex=$(RangeToRegex $1);
+
+    # Begin testing
+    for i in $(seq $from $to)
```
Comment (stelcheck) [inline regex.sh]: this does not work if from > to

mt@kodachi:~/libbash (libbash::master)$ seq 20 10
mt@kodachi:~/libbash (libbash::master)$ seq 10 20
10
11
12
13
14
15
16
17
18
19
20
mt@kodachi:~/libbash (libbash::master)$ seq 20 10
mt@kodachi:~/libbash (libbash::master)$

File: regex.sh [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+
+    # Sort from and to and get largest character length
+    if [ $from -gt $to ]
+    then
+        tmp="$from";
+        from="$to";
+        to="$tmp";
+    fi
+
+    # Ensure arguments
```
Comment (stelcheck) [inline regex.sh]: this fails

estRangeToRegex 11- 14 15; echo $?
-bash: [: 11: unary operator expected
Unkown value (11-) in range:
11-
-bash: [: 11: unary operator expected
- FAILURE on true match (11)
  0

Should be more graceful

**Comment by stelcheck** [comment]
regex.sh

I am assuming one of those commands should have returned me something, but they dont.

mt@kodachi:~/libbash (libbash::master)$ TestRangeToRegex 15-16 14 17
mt@kodachi:~/libbash (libbash::master)$ TestRangeToRegex 11-12 14 17

File: regex.sh [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+
+##
+# Takes in a comma separated list of numeric reanges which can be
+# either a single number or a range <from>-<to>. It will then
+# generate a regular expression which will match those ranges.
+#
+# Usage: RangeToRegex "<ranges>"
+#
+# Arguments:
+#     ranges: comma separated list of numeric ranges. i.e. "1,2,5,10-20,100-200"
```
Comment (stelcheck) [inline regex.sh:21]: I am assuming there should not be any final input to stdout for the following (only to stderr)

mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,100-200
1|2|5|20|1[0-9]|200|10[0-9]|1[1-9][0-9]
mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,100-  
Unkown value (100-) in range:
1,2,5,10-20,100-
mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,a100-
Unkown value (a100-) in range:
1,2,5,10-20,a100-
mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,a  
Unkown value (a) in range:
1,2,5,10-20,a
mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,a-v
Unkown value (a-v) in range:
1,2,5,10-20,a-v

**Comment by stelcheck** [comment]
Cannot test validation.sh

mt@kodachi:~/libbash (libbash::master)$ . validation.sh 
-bash: [: missing `]'

mt@kodachi:~/libbash (libbash::master)$ ./validation.sh 
./validation.sh: line 14: [: missing `]'

File: bashr.sh [hunk 9]
```diff
@@ -141,8 +141,8 @@ setup(){
         exit;
     elif [ $cmdRet -ne 0 ]; then
         echo -e "✘" | red | bold;
-        echo "-- $(cat $errorPipe) (error code: $?)" | red | bold 1>&2;
-        exit;
+        echo "-- $(cat $errorPipe) (error code: $cmdRet)" | red | bold 1>&2;
```
Comment (stelcheck) [inline bashr.sh:144]: Nice catch!

File: validation.sh [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+# description     : This script provides helper functions for data, input and
+#                   setup validation.
+# author          : Almir Kadric
+# created on      : 2012-12-12
+# version         : 0.1
+##
+
+# Get script root path, symlink resistant
+scriptRoot=$(cd "$(dirname ${BASH_SOURCE})" && pwd);
+if [ -h ${BASH_SOURCE}
```
Comment (stelcheck) [inline validation.sh]: Missing closing bracket

File: validation.sh [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+function EnsureCommandsAvailable() {
+    # Define required variables
+    commands="$@";
+    errors=false;
+
+    # Iterate through commands
+    for command in ${commands}
+    do
+        # Check command type
+        result=$(type $command);
```
Comment (stelcheck) [inline validation.sh]: should be

result=$(type $command 2>/dev/null 1>/dev/null);

File: validation.sh [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        status=$?;
+
+        # Check if type casting of command was successful
+        if [ $status -ne 0 ]
+        then
+            # Output error if command was not found
+            echoError "Could not find command ${command}";
```
Comment (stelcheck) [inline validation.sh]: I would suggest you to return a list of missing commands instead of echoing like this.

This would allow to do something like this: 

notFoundCommands=$(EnsureCommandsAvailable)

if [ "${notFoundCommands}" != "" ]; then
    echoError "the following commands are mandatory: ${notFoundCommands}"
fi

File: validation.sh [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    # Iterate through arguments
+    for argument in ${arguments}
+    do
+        label="$(echo $argument | awk -F '=' '{print $1}' | sed 's/^--//')";
+        value="$(echo $argument | awk -F '=' '{print $2}')";
+
+        if [ "$(echo ${allow} | grep ${label})" = "" ]
+        then
+            echoError "Unrecognized argument --${label}";
```
Comment (stelcheck) [inline validation.sh]: Same as for EnsureCommandsAvailable: this should output all wrong arguments instead of forcing an output format.

File: validation.sh [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+            echoError "Check --help for further usage information";
+            return 1;
```
Comment (stelcheck) [inline validation.sh]: This has the effect of showing only the first wrong flag:

ParseArguments "a b c" "d e f"
 ✘ Unrecognized argument --d
 ✘ Check --help for further usage information

File: regex.sh (hunk 2 above)
Comment (AlmirKadric) [inline regex.sh:6]: this is called regex.sh as it fits into that category, the idea is this file will hold other regex function, even though they may not exist right now

File: regex.sh (hunk 8 above)
Comment (AlmirKadric) [inline regex.sh:21]: not quite sure what you meant by this, but yes these are errors and should appear on stderr and nothing should be sent to stdout. This then will throw a non zero status which is caught on the other side, you can then decide what to do.

File: regex.sh (hunk 7 above)
Comment (AlmirKadric) [inline regex.sh]: I could not replicate the issue you mentioned.

Mac OS X:
almir-work:libbash AlmirKadric$ . ./regex.sh; RangeToRegex 11- 14 15; echo $?
Unkown value (11-) in range:
11-
1

Debian:
root@master3:/opt/Moria/lib/libbash# . ./regex.sh; RangeToRegex 11- 14 15; echo $?
Unkown value (11-) in range:
11-
1

File: regex.sh (hunk 6 above)
Comment (AlmirKadric) [inline regex.sh]: this is a non issue, check above

```
        # Sort from and to and get largest character length
        if [ $from -gt $to ]
        then
            tmp="$from";
            from="$to";
            to="$tmp";
        fi
```

File: validation.sh (hunk 10 above)
Comment (AlmirKadric) [inline validation.sh]: this was fixed, but forgot to push it

File: regex.sh (hunk 1 above)
Comment (AlmirKadric) [inline regex.sh]: ill split this off into a separate file, as for feeding the data, if you look below, that function feeds data into this function. The idea is you can quickly construct unit tests via command line and put edge cases in below function or write other function run those edge cases.

Comment (AlmirKadric) [inline regex.sh]: to simplify this push I won't overly work on the unit tests, we can change them as we need them.

File: regex.sh (hunk 2 above)
Comment (stelcheck) [inline regex.sh:6]: Fair enough

Comment (stelcheck) [inline regex.sh:6]: However, if you have a RangeToRegex function, I expect it to support alphanumerical, not only numerical then.

``` bash
mt@kodachi:~/libbash (libbash::master)$ RangeToRegex 1,2,5,10-20,a-v
Unkown value (a-v) in range:
1,2,5,10-20,a-v
```

I also would expect it to be able to deal with open lists

10,100-, etc, but that could arguably be done later

File: regex.sh (hunk 8 above)
Comment (stelcheck) [inline regex.sh:21]: Ok, good then

File: regex.sh (hunk 1 above)
Comment (stelcheck) [inline regex.sh]: What you provided is not what I have requested. 

To be more specific:

1) Unit tests shall be non blocking of one each other. I should be able to jump to the next one even if the previous does not pass.

2) They should be separate files which run the command, check the result and exit with 0 or 1. This will make it easier to work out with bashr in the future.

**Comment by stelcheck** [comment]
This issue does not look like it has been seen (see above)

https://github.com/Wizcorp/libbash/pull/2#issuecomment-12587815

File: regex.sh (hunk 2 above)
Comment (AlmirKadric) [inline regex.sh:6]: I guess I can change the name, but generally a range is numerical, I guess that was a bad assumption to make.

Should I rename the function then?

Currently only gamepad2 uses it in 1 spot so wont be too much of an impact to rename

File: validation.sh [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            then
+                echoError "Could not find command ${command}";
+            fi
+
+            notFoundCommands+=(${command});
+            errors=true;
+        fi
+    done
+
+    # Exit if any commands were not found
```
Comment (stelcheck) [inline validation.sh:74]: You seem to have changed completely your strategy here. It's a bit confusing because of the timespan between responses and because youre doing multiple code iteration within a pull request. 

Anyway, here I find that exporting is a bit... well, it leaves traces. 

I found a nice way to deal with returning values from function, which goes as following

``` bash
local resultvar=$1;
local ret;
shift;

...

eval $resultvar="'$ret'"
```

This allows for better control of where information should be returned, and does not contaminate the environment the application is running in.

Example: https://github.com/Wizcorp/libbash/blob/master/query.sh

File: regex.sh (hunk 2 above)
Comment (AlmirKadric) [inline regex.sh:6]: adding alphanumerical support is not feasible at this time

Comment (stelcheck) [inline regex.sh:6]: Not a big deal, you can leave it like this if you plan to do more with this
in the future.

On Tue, Feb 26, 2013 at 2:11 PM, Almir Kadric notifications@github.comwrote:

> In regex.sh:
> 
> > @@ -0,0 +1,453 @@
> > +#!/bin/bash
> > +
> > +##
> > +# title           : regex.sh
> > +# description     : This script provides helper functions for
> > +#                   creating or manipulating regular expressions.
> 
> I guess I can change the name, but generally a range is numerical, I guess
> that was a bad assumption to make.
> 
> Should I rename the function then?
> 
> Currently only gamepad2 uses it in 1 spot so wont be too much of an impact
> to rename
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Wizcorp/libbash/pull/2/files#r3148103
> .

## 

http://www.wizcorp.jp/Marc Trudel-Belisle

## Chief Technology Officer | Wizcorp Inc. http://www.wizcorp.jp/

TECH . GAMING . OPEN-SOURCE WIZARDS+ 81
3-4550-1448|Websitehttp://www.wizcorp.jp/
|Twitter https://twitter.com/Wizcorp|Facebookhttp://www.facebook.com/Wizcorp
|LinkedIn http://www.linkedin.com/company/wizcorp

File: validation.sh (hunk 15 above)
Comment (AlmirKadric) [inline validation.sh:74]: ah yeah, forgot about that, i was using it elsewhere but forgot about it for this implementation. Will fix it here and below

Comment (AlmirKadric) [inline validation.sh:74]: and the reason i changed my strategy is because you asked for a list to be returned, so i have to store that list and return out once all are checked as opposed to bailing on first invalidation.

Comment (stelcheck) [inline validation.sh:74]: You switched from output with style to no output at all where my suggestion was to output without style. 

You went a step further, and thats fine (this might actually be better). But now I have to look at a completely different code, which makes it insanely hard to review, and is why I would have preferred to close/open a new pull requests in such case. Keep in mind, this is a comment about the whole PR, not only about this bit of code.

Comment (stelcheck) [inline validation.sh:74]: Has this been fixed?

Comment (AlmirKadric) [inline validation.sh:74]: has what been fixed? its been awaiting your review since your above comments

Comment (stelcheck) [inline validation.sh:74]: I quote:

I found a nice way to deal with returning values from function, which goes as following

``` bash
local resultvar=$1;
local ret;
shift;

...

eval $resultvar="'$ret'"
```

This allows for better control of where information should be returned, and does not contaminate the environment the application is running in.

Example: https://github.com/Wizcorp/libbash/blob/master/query.sh

Right now you send it to output.

To this comment, you replied that you would implement such logic, but I dont see any change in the code in that regard.

Comment (AlmirKadric) [inline validation.sh:74]: Err that is inside it, thus the comment from above you sent me:

You switched from output with style to no output at all where my suggestion was to output without style.

You went a step further, and thats fine (this might actually be better). But now I have to look at a completely different code, which makes it insanely hard to review, and is why I would have preferred to close/open a new pull requests in such case. Keep in mind, this is a comment about the whole PR, not only about this bit of code.

As you can see in the code I have:
eval notFoundCommands+=(${command});

and the make silent var around the actual echo:
if [ "$SILENT_ERROR" = "" ] || ! $SILENT_ERROR

This way you can have an error to stdout by default if you're quickly whipping something up. But once you get more custom, you can silent it and do a custom output based off of notFoundCommands.

Comment (stelcheck) [inline validation.sh:74]: I understand that, but what I was suggesting is to pass the variable name as argument. Else, I am not sure why we would even need to eval.

Comment (AlmirKadric) [inline validation.sh:74]: I updated the libbash pull request for both function to follow the logic we discussed
we will output missing/invlid parts to stdout and output error code upon completion
so a script will use it in this manner

<pre>
# Check that all required commmands are installed
if ! missingCMDs=$(EnsureCommandsAvailable "s3cmd mysql mysqldump innobackupex hostname mkdir rm mv sed")
then
    echo "The following system commands could not be found:";
    echo "    $(echo ${missingCMDs[@]} | sed 's/^/--/g ; s/\ /\ --/g')";
    echo "";
    exit 1;
fi
</pre>

Comment (stelcheck) [inline validation.sh:74]: sounds good. merging.
